From: Vishvananda Ishaya <vishvananda@gmail.com>
Date: Thu, 1 Mar 2012 16:52:07 -0800
Subject: [PATCH] Fixes for ec2 images

 * Fixes s3 image service to convert back to uuids on update
 * Adds exception for attempt to update an unowned image
 * Adds error messages to ec2 for failure cases
 * Adds tests to verify changes
 * Fixes bug 942865

Change-Id: I35331c635756f10c02b30dd43ab3fe0ad98bc28c

Origin: https://review.openstack.org/#change,4788
Bug: https://code.launchpad.net/bugs/942865
Last-Updated: 2012-03-05

Index: nova/nova/api/ec2/cloud.py
===================================================================
--- nova.orig/nova/api/ec2/cloud.py	2012-03-05 14:27:39.000000000 -0800
+++ nova/nova/api/ec2/cloud.py	2012-03-05 14:28:45.223610255 -0800
@@ -1434,6 +1434,9 @@
     def register_image(self, context, image_location=None, **kwargs):
         if image_location is None and kwargs.get('name'):
             image_location = kwargs['name']
+        if image_location is None:
+            raise exception.EC2APIError(_('imageLocation is required'))
+
         metadata = {'properties': {'image_location': image_location}}
 
         if kwargs.get('name'):
@@ -1513,7 +1516,11 @@
         del(image['id'])
 
         image['is_public'] = (operation_type == 'add')
-        return self.image_service.update(context, internal_id, image)
+        try:
+            return self.image_service.update(context, internal_id, image)
+        except exception.ImageNotAuthorized:
+            msg = _('Not allowed to modify attributes for image %s')
+            raise exception.EC2APIError(msg % image_id)
 
     def update_image(self, context, image_id, **kwargs):
         internal_id = ec2utils.ec2_id_to_id(image_id)
Index: nova/nova/exception.py
===================================================================
--- nova.orig/nova/exception.py	2012-03-05 14:27:39.000000000 -0800
+++ nova/nova/exception.py	2012-03-05 14:28:45.223610255 -0800
@@ -201,7 +201,7 @@
 
 class NotAuthorized(NovaException):
     message = _("Not authorized.")
-    code = 401
+    code = 403
 
 
 class AdminRequired(NotAuthorized):
@@ -212,6 +212,10 @@
     message = _("Policy doesn't allow %(action)s to be performed.")
 
 
+class ImageNotAuthorized(NovaException):
+    message = _("Not authorized for image %(image_id)s.")
+
+
 class Invalid(NovaException):
     message = _("Unacceptable parameters.")
     code = 400
Index: nova/nova/image/glance.py
===================================================================
--- nova.orig/nova/image/glance.py	2012-03-05 14:27:39.000000000 -0800
+++ nova/nova/image/glance.py	2012-03-05 14:28:45.223610255 -0800
@@ -301,6 +301,9 @@
             image_meta = client.update_image(image_id, image_meta, data)
         except glance_exception.NotFound:
             raise exception.ImageNotFound(image_id=image_id)
+        # NOTE(vish): this gets raised for public images
+        except glance_exception.MissingCredentialError:
+            raise exception.ImageNotAuthorized(image_id=image_id)
 
         base_image_meta = self._translate_from_glance(image_meta)
         return base_image_meta
Index: nova/nova/image/s3.py
===================================================================
--- nova.orig/nova/image/s3.py	2012-03-05 14:27:39.000000000 -0800
+++ nova/nova/image/s3.py	2012-03-05 14:28:45.223610255 -0800
@@ -105,6 +105,27 @@
 
         return image_copy
 
+    def _translate_id_to_uuid(self, context, image):
+        image_copy = image.copy()
+
+        try:
+            image_id = image_copy['id']
+        except KeyError:
+            pass
+        else:
+            image_copy['id'] = self.get_image_uuid(context, image_id)
+
+        for prop in ['kernel_id', 'ramdisk_id']:
+            try:
+                image_id = image_copy['properties'][prop]
+            except (KeyError, ValueError):
+                pass
+            else:
+                image_uuid = self.get_image_uuid(context, image_id)
+                image_copy['properties'][prop] = image_uuid
+
+        return image_copy
+
     def create(self, context, metadata, data=None):
         """Create an image.
 
@@ -120,6 +141,7 @@
 
     def update(self, context, image_id, metadata, data=None):
         image_uuid = self.get_image_uuid(context, image_id)
+        metadata = self._translate_id_to_uuid(context, metadata)
         image = self.service.update(context, image_uuid, metadata, data)
         return self._translate_uuid_to_id(context, image)
 
@@ -171,6 +193,11 @@
         image_type = 'machine'
 
         try:
+            image_name = manifest.find('image/name').text
+        except Exception:
+            image_name = None
+
+        try:
             kernel_id = manifest.find('machine_configuration/kernel_id').text
             if kernel_id == 'true':
                 image_format = 'aki'
@@ -235,6 +262,7 @@
 
         metadata.update({'disk_format': image_format,
                          'container_format': image_format,
+                         'name': image_name,
                          'status': 'queued',
                          'is_public': False,
                          'properties': properties})
Index: nova/nova/tests/api/ec2/test_cloud.py
===================================================================
--- nova.orig/nova/tests/api/ec2/test_cloud.py	2012-03-05 14:27:39.287612677 -0800
+++ nova/nova/tests/api/ec2/test_cloud.py	2012-03-05 14:28:45.223610255 -0800
@@ -1156,7 +1156,7 @@
         modify_image_attribute = self.cloud.modify_image_attribute
 
         fake_metadata = {
-            'id': 1,
+            'id': 'cedef40a-ed67-4d10-800e-17455edce175',
             'container_format': 'ami',
             'properties': {
                 'kernel_id': 'cedef40a-ed67-4d10-800e-17455edce175',
@@ -1165,11 +1165,17 @@
             'is_public': False}
 
         def fake_show(meh, context, id):
-            return fake_metadata
+            return copy.deepcopy(fake_metadata)
 
         def fake_update(meh, context, image_id, metadata, data=None):
-            fake_metadata.update(metadata)
-            return fake_metadata
+            self.assertEqual(metadata['properties']['kernel_id'],
+                             fake_metadata['properties']['kernel_id'])
+            self.assertEqual(metadata['properties']['ramdisk_id'],
+                             fake_metadata['properties']['ramdisk_id'])
+            self.assertTrue(metadata['is_public'])
+            image = copy.deepcopy(fake_metadata)
+            image.update(metadata)
+            return image
 
         self.stubs.Set(fake._FakeImageService, 'show', fake_show)
         self.stubs.Set(fake._FakeImageService, 'show_by_name', fake_show)
@@ -1177,7 +1183,31 @@
         result = modify_image_attribute(self.context, 'ami-00000001',
                                           'launchPermission', 'add',
                                            user_group=['all'])
-        self.assertEqual(True, result['is_public'])
+        self.assertTrue(result['is_public'])
+
+    def test_register_image(self):
+        register_image = self.cloud.register_image
+
+        def fake_create(*args, **kwargs):
+            # NOTE(vish): We are mocking s3 so make sure we have converted
+            #             to ids instead of uuids.
+            return {'id': 1,
+            'container_format': 'ami',
+            'properties': {
+                'kernel_id': 1,
+                'ramdisk_id': 1,
+                'type': 'machine'},
+            'is_public': False}
+
+        self.stubs.Set(s3.S3ImageService, 'create', fake_create)
+        image_location = 'fake_bucket/fake.img.manifest.xml'
+        result = register_image(self.context, image_location)
+        self.assertEqual(result['imageId'], 'ami-00000001')
+
+    def test_register_image_empty(self):
+        register_image = self.cloud.register_image
+        self.assertRaises(exception.EC2APIError, register_image, self.context,
+                          image_location=None)
 
     def test_register_image_name(self):
         register_image = self.cloud.register_image
